src/macaron/slsa_analyzer/analyzer.py

```diff
@@ -540,6 +541,14 @@ def to_analysis_target(config: Configuration, available_domains: list[str]) -> A
             except ValueError as error:
                 raise InvalidPURLError(f"Invalid input PURL: {config.get_value('purl')}") from error
 
+        # Validate the purl object, which is user-controllable.
+        if parsed_purl:
+            purl_str = parsed_purl.to_string()
+            # The urlparse API sanitizes unsafe characters.
+            parsed_url = urllib.parse.urlparse(purl_str)
+            if purl_str != parsed_url.geturl():
+                raise ValueError("The PURL provided as input has unsafe characters.")
```

Comment:
Because this seems important and probably can be tested, I wonder if we can extract this into a function and unit-test it.

Reply:
I agree.